cloud/scope/powervs_cluster_test.go

```diff
@@ -3043,3 +3043,2423 @@ func TestCreateVPCSubnet(t *testing.T) {
+					MinimumPort: 1,
+				},
+			},
+		}
+
+		vpcSecurityGroup := infrav1beta2.VPCSecurityGroup{
+			Name:  ptr.To("securityGroupName"),
+			Rules: append([]*infrav1beta2.VPCSecurityGroupRule{}, &rules),
+		}
+		clusterScope := PowerVSClusterScope{
+			IBMVPCClient: mockVPC,
+			IBMPowerVSCluster: &infrav1beta2.IBMPowerVSCluster{
+				Status: infrav1beta2.IBMPowerVSClusterStatus{
+					VPC: &infrav1beta2.ResourceReference{
+						ID: ptr.To("VPCID"),
+					},
+				},
+				Spec: infrav1beta2.IBMPowerVSClusterSpec{
+					VPCSecurityGroups: append([]infrav1beta2.VPCSecurityGroup{}, vpcSecurityGroup),
+				},
+			},
+		}
+
+		mockVPC.EXPECT().CreateSecurityGroupRule(gomock.Any()).Return(&vpcv1.SecurityGroupRuleSecurityGroupRuleProtocolAll{Direction: ptr.To("inbound"), ID: ptr.To("ruleID")}, nil, nil)
+		ruleIDs, err := clusterScope.createVPCSecurityGroupRules(vpcSecurityGroup.Rules, ptr.To("securityGroupID"))
+		g.Expect(ruleIDs).To(BeEquivalentTo([]*string{ptr.To("ruleID")}))
+		g.Expect(err).To(BeNil())
+	})
+	t.Run("When SecurityGroupRule of Direction Inbound returns error", func(t *testing.T) {
+		g := NewWithT(t)
+		setup(t)
+		t.Cleanup(teardown)
+		rules := infrav1beta2.VPCSecurityGroupRule{
+			Direction: infrav1beta2.VPCSecurityGroupRuleDirectionInbound,
+			Source: &infrav1beta2.VPCSecurityGroupRulePrototype{
+				Remotes: append([]infrav1beta2.VPCSecurityGroupRuleRemote{}, infrav1beta2.VPCSecurityGroupRuleRemote{
+					CIDRSubnetName: ptr.To("CIDRSubnetName"),
+					RemoteType:     infrav1beta2.VPCSecurityGroupRuleRemoteTypeCIDR,
+				}),
+				Protocol: infrav1beta2.VPCSecurityGroupRuleProtocolTCP,
+				PortRange: &infrav1beta2.VPCSecurityGroupPortRange{MaximumPort: 65535,
+					MinimumPort: 1,
+				},
+			},
+		}
+
+		vpcSecurityGroup := infrav1beta2.VPCSecurityGroup{
+			Name:  ptr.To("securityGroupName"),
+			Rules: append([]*infrav1beta2.VPCSecurityGroupRule{}, &rules),
+		}
+		clusterScope := PowerVSClusterScope{
+			IBMVPCClient: mockVPC,
+			IBMPowerVSCluster: &infrav1beta2.IBMPowerVSCluster{
+				Status: infrav1beta2.IBMPowerVSClusterStatus{
+					VPC: &infrav1beta2.ResourceReference{
+						ID: ptr.To("VPCID"),
+					},
+				},
+				Spec: infrav1beta2.IBMPowerVSClusterSpec{
+					VPCSecurityGroups: append([]infrav1beta2.VPCSecurityGroup{}, vpcSecurityGroup),
+				},
+			},
+		}
+
+		mockVPC.EXPECT().GetVPCSubnetByName(gomock.Any()).Return(nil, nil)
+		ruleIDs, err := clusterScope.createVPCSecurityGroupRules(vpcSecurityGroup.Rules, ptr.To("securityGroupID"))
+		g.Expect(ruleIDs).To(BeNil())
+		g.Expect(err).ToNot(BeNil())
+	})
+}
+
+func TestCreateVPCSecurityGroupRulesAndSetStatus(t *testing.T) {
+	var (
+		mockVPC  *mock.MockVpc
+		mockCtrl *gomock.Controller
+	)
+	setup := func(t *testing.T) {
+		t.Helper()
+		mockCtrl = gomock.NewController(t)
+		mockVPC = mock.NewMockVpc(mockCtrl)
+	}
+	teardown := func() {
+		mockCtrl.Finish()
+	}
+	t.Run("When SecurityGroupRule is created successfully", func(t *testing.T) {
```

Comment:
Another one for `createVPCSecurityGroupRules` erroring out?

Reply:
added it 
https://github.com/kubernetes-sigs/cluster-api-provider-ibmcloud/pull/2011/files#diff-04d9f443c4026a9cd5e26cf630140aa890cf62140d697655d0293a8d97970f57R5401